src/coap_gnutls.c

```diff
@@ -0,0 +1,1691 @@
+        name = new_name;
+        continue; /* retry call with same index */
+      }
+
+      /* check if it is the last entry in list */
+      if (ret == GNUTLS_E_REQUESTED_DATA_NOT_AVAILABLE)
+        break;
+      i++;
+      if (ret != GNUTLS_E_SUCCESS)
+        goto end;
+      /* unknown types need to be ignored */
+      if (type != GNUTLS_NAME_DNS)
+        continue;
+
+    }
+    /* If no extension provided, make it a dummy entry */
+    if (i == 0) {
+      name[0] = '\000';
+      len = 0;
+    }
+
+    /* Is this a cached entry? */
+    for (i = 0; i < g_context->sni_count; i++) {
+      if (strcmp(name, g_context->sni_entry_list[i].sni) == 0) {
+        break;
+      }
+    }
+    if (i == g_context->sni_count) {
+      /*
+       * New SNI request
+       */
+      coap_dtls_key_t *new_entry =
+        g_context->setup_data.validate_sni_call_back(name,
+                                   g_context->setup_data.sni_call_back_arg);
+      if (!new_entry) {
+        G_ACTION(gnutls_alert_send(g_session, GNUTLS_AL_FATAL,
+                                   GNUTLS_A_UNRECOGNIZED_NAME));
+        ret = GNUTLS_E_NO_CERTIFICATE_FOUND;
+        goto end;
+      }
+
+      g_context->sni_entry_list = gnutls_realloc(g_context->sni_entry_list,
+                                     (i+1)*sizeof(sni_entry));
+      g_context->sni_entry_list[i].sni = gnutls_strdup(name);
+      g_context->sni_entry_list[i].pki_key = *new_entry;
+      sni_setup_data = g_context->setup_data;
+      sni_setup_data.pki_key = *new_entry;
+      if ((ret = setup_pki_credentials(
+                           &g_context->sni_entry_list[i].pki_credentials,
+                           g_context,
+                           &sni_setup_data)) < 0) {
+        int keep_ret = ret;
+        G_ACTION(gnutls_alert_send(g_session, GNUTLS_AL_FATAL,
+                                   GNUTLS_A_BAD_CERTIFICATE));
+        ret = keep_ret;
+        goto end;
+      }
+      g_context->sni_count++;
+    }
+    G_CHECK(gnutls_credentials_set(g_env->g_session, GNUTLS_CRD_CERTIFICATE,
+                               g_context->sni_entry_list[i].pki_credentials),
+            "gnutls_credentials_set");
+  }
+
+end:
+  free(name);
+  return ret;
+
+fail:
+  return ret;
+}
+
+/*
+ * return 0   Success (GNUTLS_E_SUCCESS)
+ *        neg GNUTLS_E_* error code
+ */
+static int
+setup_client_ssl_session(coap_session_t *c_session, coap_gnutls_env_t *g_env)
+{
+  coap_gnutls_context_t *g_context =
+             (coap_gnutls_context_t *)c_session->context->dtls_context;
+  int ret;
+
+  g_context->psk_pki_enabled |= IS_CLIENT;
+  if (g_context->psk_pki_enabled & IS_PSK) {
+    char *identity;
+    gnutls_datum_t psk_key;
+
+    G_CHECK(gnutls_psk_allocate_client_credentials(&g_env->psk_cl_credentials),
+            "gnutls_psk_allocate_client_credentials");
+    psk_client_callback(g_env->g_session, &identity, &psk_key);
+    G_CHECK(gnutls_psk_set_client_credentials(g_env->psk_cl_credentials,
```

Comment:
```sh
src/coap_gnutls.c: In function ‘setup_client_ssl_session.isra.6’:
src/coap_gnutls.c:850:13: warning: ‘identity’ may be used uninitialized in this function [-Wmaybe-uninitialized]
     G_CHECK(gnutls_psk_set_client_credentials(g_env->psk_cl_credentials,
             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Reply:
Fixed (all above missed as I was compiling with -ggdb3 locally for debugging the code - sorry).